### PR TITLE
br: add GetTSWithRetry func

### LINKED ISSUE #36910: BR：When the abnormal switch of the leader occurs in the pd, it should not affect the br backup process
version 5.4.1
I encounter a case: when pd OOM and switch leader, the br will fail.

### PR DESCRIPTION
**Pull request by MoCuishle28**
### What problem does this PR solve?

Issue Number: ref #36910

Problem Summary:
When the switch of the leader occurs in the PD, it affect the BR restore checksum process.

### What is changed and how it works?
Add `GetTSWithRetry` func to replace `GetTS` in `execChecksum` func

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Manual Test
- run restore with injected error
- restore succeeded

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix a bug that caused: When the switch of the leader occurs in the PD, it affect the BR restore checksum process.
```

### CODE REVIEW

File: br/pkg/restore/client.go [hunk 1]
```diff
@@ -494,6 +494,32 @@ func (rc *Client) GetTS(ctx context.Context) (uint64, error) {
 	return restoreTS, nil
 }
 
+// GetTSWithRetry gets a new timestamp with retry from PD.
+func (rc *Client) GetTSWithRetry(ctx context.Context) (uint64, error) {
+	var startTS uint64
+	var getTSErr error = nil
+	var retry uint = 0
```
Comment (YuJuncen) [inline br/pkg/restore/client.go:501]: Perhaps we can emit zero values and merge those `var`s:
```suggestion
	var (
		startTS uint64
		getTSErr error
		retry uint
	)
```

File: br/pkg/restore/client_test.go [hunk 2]
```diff
@@ -330,12 +331,35 @@ func TestPreCheckTableClusterIndex(t *testing.T) {
...
 
+func (fpdc fakePDClient) GetTS(ctx context.Context) (int64, int64, error) {
+	if fpdc.notLeader {
+		return 0, 0, errors.Errorf("rpc error: code = Unknown desc = [PD:tso:ErrGenerateTimestamp]generate timestamp failed, requested pd is not leader of cluster")
+	}
+	return 1, 1, nil
+}
+
+func TestGetTSWithRetry(t *testing.T) {
+	t.Log("case 1:")
```
Comment (YuJuncen) [inline br/pkg/restore/client_test.go:350]: Perhaps using [`t.Run`](https://pkg.go.dev/testing@go1.19.2#T.Run) to start sub-cases.

File: br/pkg/restore/client_test.go [hunk 3]
```diff
@@ -330,12 +331,35 @@ func TestPreCheckTableClusterIndex(t *testing.T) {
...
+	pDClient := fakePDClient{notLeader: false}
+	client := restore.NewRestoreClient(pDClient, nil, defaultKeepaliveCfg, false)
+	_, err := client.GetTSWithRetry(context.Background())
+	require.NoError(t, err)
+
+	t.Log("case 2:")
+	pDClient = fakePDClient{notLeader: true}
+	client = restore.NewRestoreClient(pDClient, nil, defaultKeepaliveCfg, false)
+	_, err = client.GetTSWithRetry(context.Background())
+	require.Error(t, err)
```
Comment (YuJuncen) [inline br/pkg/restore/client_test.go:360]: Can we add a case that the mock PD fails in the first request, then success at the second try? (Which should also return no errors.)

**Review by YuJuncen — COMMENTED** [review]
rest lgtm

File: br/pkg/restore/client.go [hunk 4]
```diff
@@ -1374,7 +1400,7 @@ func (rc *Client) execChecksum(
 		ctx = opentracing.ContextWithSpan(ctx, span1)
 	}
 
-	startTS, err := rc.GetTS(ctx)
+	startTS, err := rc.GetTSWithRetry(ctx)
```
Comment (3pointer) [inline br/pkg/restore/client.go:1403]: can we replace all `rc.GetTS` with `rc.GetTSWithRetry`?

File: br/pkg/restore/client_test.go (hunk 3 above)
Comment (MoCuishle28) [reply br/pkg/restore/client_test.go:360]: I can simulate this case. But maybe this case is already covered when `get-ts-error` failpoint enable?

File: br/pkg/restore/client.go (hunk 4 above)
Comment (MoCuishle28) [reply br/pkg/restore/client.go:1403]: Maybe yes. But `GetTS` is widely used, so I think it needs careful evaluation and testing.

File: br/pkg/restore/client_test.go (hunk 3 above)
Comment (MoCuishle28) [reply br/pkg/restore/client_test.go]: I added a new unit test case to simulate this scenario. PTAL

**Review by YuJuncen — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]

**Pull request merged**